### PR TITLE
Add `MultinomialRV` JAX implementation

### PR DESCRIPTION
**Pull request by GStechschulte**
This draft PR is a work in progress and contains a JAX implementation of `MultinomialRV` for issue #1326. The implementation builds off the `Multinomial Distribution` [implementation](https://num.pyro.ai/en/v0.2.4/_modules/numpyro/distributions/util.html) in NumPyro. Likewise, the output is similar to that of the `numpy` [implementation](https://numpy.org/doc/stable/reference/random/generated/numpy.random.multinomial.html).  Below, you will find a brief outline of the functions used to construct the `MultinomialRV`. 

`def _categorical(key, p, shape)`

- returns the outcomes $k$ with probability $p$ for each trial / experiment $n$. 

`def _scatter_add_ones(operand, indices, update)` 

- returns the outcome counts by utilising the `jax.lax.scatter_add()` function
- `operand` is a zero filled array. 
- `indices` is the `outcomes` array with an added dimension and specifies the indices to which the update should be applied to. 
- `update` is an array filled with ones and can be thought of as a `cnt += 1` for each $K = k$ occurrence.
- In summary, the `operand` array is updated `+1` using the `update` array according to the outcomes in the `indices` array.

I still need to add a test for this. Thanks!

### CODE REVIEW

**Comment by rlouf** [comment]
Thanks! We also need to figure out if the licenses are compatible and how to do proper attribution if you took inspiration from someone else's implementation

**Comment by GStechschulte** [comment]
> Thanks! We also need to figure out if the licenses are compatible and how to do proper attribution if you took inspiration from someone else's implementation. It looks like [Numpyro is licensed under Apache 2.0](https://github.com/pyro-ppl/numpyro/blob/master/LICENSE.md)

Based on the NumPyro Apache License 2.0 section 4, we may reproduce and distribute copies of the Work or Derivative Works (the JAX implementation of MultinomialRV) provided we:

1. give any other recipients of the Work or Derivative Works a copy of this License; and
2. the modified file must contain carry a notice stating the file was changed
3. in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work
4. if the Work includes a "NOTICE" text file as part of its distribution, then any Derivative Works that You distribute must include a readable copy of the attribution notices contained within such NOTICE file, excluding those notices that do not pertain to any part of the Derivative Works, in at least one of the following places: within a NOTICE text file distributed as part of the Derivative Works; within the Source form or documentation

Since inspiration was drawn from a few functions and not an entire file, I suggest we include, in addition to (1), (2), (3), in the documentation for this RV, something along the lines 

> MultinomialRV uses source code from the file xyz.py from <link to src code GitHub file of the NumPyro project, copyright YYYY, licensed under the Apache 2.0 license>

File: aesara/link/jax/dispatch/random.py [hunk 1]
```diff
@@ -277,3 +277,49 @@ def sample_fn(rng, size, dtype, *parameters):
...
+            )
+
+        return outcome_cnts
+
+    def _categorical(key, p, shape):
+        shape = shape or p.shape[:-1]
+        s = jax.numpy.cumsum(p, axis=-1)
+        r = jax.random.uniform(key, shape=shape + (1,))
+        
+        return jax.numpy.sum(s < r, axis=-1)
```
Comment (rlouf) [inline aesara/link/jax/dispatch/random.py:303]: I am surprised because JAX does have an implementation for the categorical distribution [here](https://jax.readthedocs.io/en/latest/_autosummary/jax.random.categorical.html#jax.random.categorical) that uses their implementation for the Gumbel distribution. Is this justified in the codebase, or is it just because it was implemented at a time where `jax.random.categorical` was not available (which you should be able to determine with `git blame`)?

File: aesara/link/jax/dispatch/random.py [hunk 2]
```diff
@@ -277,3 +277,49 @@ def sample_fn(rng, size, dtype, *parameters):
...
+        
+        outcomes = _categorical(rng, p, (n_max,) + size)
+        outcomes_2d = (jax.numpy.reshape(outcomes, (n_max, -1,))).T
+        samples_2d = jax.vmap(_scatter_add_one, (0, 0, 0))(
+            jax.numpy.zeros((outcomes_2d.shape[0], p.shape[-1]), dtype=outcomes.dtype),
+            jax.numpy.expand_dims(outcomes_2d, axis=-1),
+            jax.numpy.ones(outcomes_2d.shape, dtype=outcomes.dtype)
+            )
+        
+        sample = jax.numpy.reshape(samples_2d, size + p.shape[-1:])
```
Comment (rlouf) [inline aesara/link/jax/dispatch/random.py:320]: Couldn't we use `jax.nn.one_hot` on the output of the categorical and then reduce the resulting tensor?

**Comment by rlouf** [comment]
I rebased your branch on `main` to use the new key splitting scheme in the JAX backend. You'll have to pull the changes!

**Review by GStechschulte — COMMENTED** [review]
When I performed the `git pull --merge`, it kept a copy of the previous code (without the new key splitting). This was likely a mistake on my end. Therefore, I deleted it.

**Comment by rlouf** [comment]
Yes you need to `git pull --rebase` in such cases. Here's [a good explanation of how it works](https://git-scm.com/docs/git-rebase).

File: aesara/link/jax/dispatch/random.py [hunk 3]
```diff
@@ -326,3 +326,34 @@ def sample_fn(rng, size, dtype, *parameters):
...
+@jax_sample_fn.register(aer.MultinomialRV)
+def jax_sample_fn_multinomial(op):
+    """JAX implementation of `MultinomialRV`."""
+
+    def _categorical(key, p, shape):
+        shape = shape or p.shape[:-1]
+        s = jax.numpy.cumsum(p, axis=-1)
+        r = jax.random.uniform(key, shape=shape + (1,))
+
+        return jax.numpy.sum(s < r, axis=-1)
```
Comment (rlouf) [inline aesara/link/jax/dispatch/random.py:340]: Is that different from `jax.random.categorical`?

Comment (GStechschulte) [reply aesara/link/jax/dispatch/random.py:340]: Yes, the `jax.random.categorical` uses their implementation of the Gumbel distribution. However, after using `git blame`, it seems the NumPyro team used this implementation before the implementation of the `jax.random.categorical`. 

If we would like to update this code to use the `jax.random.categorical`, it would be the following:

```python3

def sample_fn(rng, size, dtype, *parameters):

        rng_key = rng["jax_state"]
        rng_key, sampling_key = jax.random.split(rng_key, 2)

        n, p = parameters
        n_max = jax.numpy.max(n)
        size = size or p.shape[:-1]
    
        logits = jax.scipy.special.logit(p)
        indices = jax.random.categorical(jax_key, logits, shape=(n_max,) + size)
        one_hot = jax.nn.one_hot(indices, p.shape[0])
        sample = jax.numpy.sum(one_hot, axis=0, dtype=dtype, keepdims=False)

        rng["jax_state"] = rng_key

        return (rng, sample)
```

Comment (rlouf) [reply aesara/link/jax/dispatch/random.py:340]: That looks much simpler, great!

Since the multinomial distribution is slightly more complex than the other ones when it comes to shapes we should make sure that the output shape of the samples that are generated in the JAX backend is identical to those of the samples generated with the other backends.

Comment (AdrienCorenflos) [reply aesara/link/jax/dispatch/random.py]: I tend to use `jax.random.choice` for this kind of thing. jax.random.categorical (via Gumbel) has a quadratic complexity.

image

Comment (AdrienCorenflos) [reply aesara/link/jax/dispatch/random.py]: You just need to check the Gumbel implementation, they form a N by N array.

File: .gitignore [hunk 4]
```diff
@@ -55,3 +55,4 @@ aesara-venv/
 testing-report.html
 coverage.xml
 .coverage.*
+jax_multinomial_test.py
```
Comment (brandonwillard) [inline .gitignore:58]: This looks like a job for a "local" Git ignore (see [here](https://stackoverflow.com/a/1753078/3006474)).

**Comment by rlouf** [comment]
@GStechschulte I think we should follow @AdrienCorenflos's suggestion here. I'll take another look at it this week.

**Comment by rlouf** [comment]
@GStechschulte I rebased your branch on `main`. Do  you plan on implementing @AdrienCorenflos's suggestion above?

**Comment by rlouf** [comment]
> @rlouf, I had to add a case for `Constant`s in `assert_size_argument_jax_compatible`. You'll need to confirm that this is valid more generally.

That's valid. I was so focused on the complex case that I forgot the simplest one.
